swaynagbar/render.c

```diff
@@ -0,0 +1,298 @@
+		return ideal_surface_height;
+	}
+
+	button->x = *x - border - text_width - padding * 2;
+	button->y = (int)(ideal_height - text_height) / 2 - padding;
+	button->width = text_width + padding * 2;
+	button->height = text_height + padding * 2;
+
+	cairo_set_source_u32(cairo, nagbar->colors.border);
+	cairo_rectangle(cairo, button->x - border, button->y - border,
+			button->width + border * 2, button->height + border * 2);
+	cairo_fill(cairo);
+
+	cairo_set_source_u32(cairo, nagbar->colors.button_background);
+	cairo_rectangle(cairo, button->x, button->y,
+			button->width, button->height);
+	cairo_fill(cairo);
+
+	cairo_set_source_u32(cairo, nagbar->colors.text);
+	cairo_move_to(cairo, button->x + padding, button->y + padding);
+	pango_printf(cairo, nagbar->font, nagbar->scale, true, "%s", button->text);
+
+	*x = button->x - border;
+
+	return ideal_height;
+}
+
+static uint32_t render_to_cairo(cairo_t *cairo, struct sway_nagbar *nagbar) {
+	uint32_t max_height = 0;
+
+	cairo_set_operator(cairo, CAIRO_OPERATOR_SOURCE);
+	cairo_set_source_u32(cairo, nagbar->colors.background);
+	cairo_paint(cairo);
+
+	uint32_t h = render_message(cairo, nagbar);
+	max_height = h > max_height ? h : max_height;
+
+	int x = (nagbar->width - NAGBAR_BUTTON_MARGIN_RIGHT) * nagbar->scale;
+	for (int i = 0; i < nagbar->buttons->length; i++) {
+		h = render_button(cairo, nagbar, i, &x);
+		max_height = h > max_height ? h : max_height;
+		x -= NAGBAR_BUTTON_GAP * nagbar->scale;
+		if (i == 0) {
+			x -= NAGBAR_BUTTON_GAP_CLOSE * nagbar->scale;
+		}
+	}
+
+	if (nagbar->details.visible) {
+		h = render_detailed(cairo, nagbar, max_height);
+		max_height = h > max_height ? h : max_height;
+	}
+
+	int border = NAGBAR_BAR_BORDER_THICKNESS * nagbar->scale;
+	if (max_height > nagbar->height) {
+		max_height += border;
+	}
+	cairo_set_source_u32(cairo, nagbar->colors.border_bottom);
+	cairo_rectangle(cairo, 0, nagbar->height * nagbar->scale - border,
+			nagbar->width * nagbar->scale, border);
+	cairo_fill(cairo);
+
+	return max_height;
+}
+
+void render_frame(struct sway_nagbar *nagbar) {
+	if (!nagbar->run_display) {
+		return;
+	}
+
+	cairo_surface_t *recorder = cairo_recording_surface_create(
+			CAIRO_CONTENT_COLOR_ALPHA, NULL);
+	cairo_t *cairo = cairo_create(recorder);
+	cairo_save(cairo);
+	cairo_set_operator(cairo, CAIRO_OPERATOR_CLEAR);
+	cairo_paint(cairo);
+	cairo_restore(cairo);
+	uint32_t height = render_to_cairo(cairo, nagbar);
+	if (height != nagbar->height) {
+		zwlr_layer_surface_v1_set_size(nagbar->layer_surface, 0, height);
+		zwlr_layer_surface_v1_set_exclusive_zone(nagbar->layer_surface,
+				height);
+		wl_surface_commit(nagbar->surface);
+		wl_display_roundtrip(nagbar->display);
+	} else {
+		nagbar->current_buffer = get_next_buffer(nagbar->shm,
+				nagbar->buffers,
+				nagbar->width * nagbar->scale,
+				nagbar->height * nagbar->scale);
+		if (!nagbar->current_buffer) {
+			wlr_log(WLR_DEBUG, "Failed to get buffer. Skipping frame.");
+			return;
```

Comment:
Surely you need to destroy the cairo surface and the cairo context here, or `goto` the end so it can be cleaned up there?